### PR TITLE
common: Add flexibility to ofi_str_toaddr()

### PR DESCRIPTION
**Pull request by j-xiong**
Allow the input address string be host name, ip address, or network interface
name in addition to the original "fi_addr[...]://address" format.

Signed-off-by: Jianxin Xiong <jianxin.xiong@intel.com>

### CODE REVIEW

**Comment by j-xiong** [comment]
@a-ilango @shefty This allows mrail address format such as FI_OFI_MRAIL_ADDR_STRC=ib0,ib1,ib2,ib3 which is not only much shorter, but also can be used at each side. One question is: do we want this capability in the common code (ofi_str_toaddr) instead?

**Comment by shefty** [comment]
I think this would be a nice addition to the common code.

**Comment by a-ilango** [comment]
👍

Should we change the name of env variable `FI_OFI_MRAIL_ADDR_STRC` to reflect that it can accept other formats as well?

**Comment by j-xiong** [comment]
@shefty I will update the PR to move the addition to the common code
@a-ilango Not necessary especially after the changes are in the common code

File: src/common.c [hunk 1]
```diff
@@ -538,12 +539,69 @@ static int ofi_str_to_sin6(const char *str, void **addr, size_t *len)
...
+		err = -FI_ENOMEM;
+		goto out;
+	}
+
+	memcpy(*addr, &ifr.ifr_addr, *len);
+
+out:
+	close(sock_fd);
+	return err;
+}
```
Comment (shefty) [inline src/common.c]: There's existing code, ofi_getifaddrs(), that's OS portable.  Can this be used instead?

Comment (j-xiong) [reply src/common.c]: That can be used. The catch is that getifaddrs is not always present (it's guarded by a configure generated definition). I think I can use ofi_getifaddrs whenever possible and use the current code as a fallback.

Comment (shefty) [reply src/common.c]: It's guarded mainly because the calls are not available on windows.  See the appveyor failures.

**Comment by shefty** [comment]
@aingerson - please see CI failures:
```
- name:   fi_av_test -g n25-mlx5_0_1 -n 1 -s n25-mlx5_0_1 -p "verbs;ofi_rxd"
  result: Fail
  time:   1
  server_cmd: /home/build/buildbot-upstreambuild_install/ofi/ofi_rhel7-dl/06/bin/fi_av_test -g n25-mlx5_0_1 -n 1 -s n25-mlx5_0_1 -p "verbs;ofi_rxd"
  server_stdout: |
    stty: standard input: Inappropriate ioctl for device
    Testing AVs on fabric IB-0xfe80000000000000
    
    Testing with type = FI_AV_MAP
    Running av_open_close [Test open and close AVs of varying sizes]...PASS!
    Running av_good_sync [Test sync AV insert with good address]...PASS!
    Running av_null_fi_addr [Test AV insert without specifying fi_addr]...skipped because: test not valid for AV type FI_AV_MAP
    Running av_good_vector_async [Test async AV insert with vector of good addresses]...PASS!
    Running av_zero_async [Test async insert AV insert of zero addresses]...PASS!
    Running av_good_2vector_async [Test async AV inserts with two address vectors]...PASS!
    
    Testing with invalid address
    Running av_bad_sync [Test sync AV insert of bad address]...FAIL: fi_av_insert ret=1, should be 0
    Running av_goodbad_vector_sync [Test sync AV insert of 1 good and 1 bad address]...FAIL: fi_av_insert ret=0, should be 1
    Running av_goodbad_vector_async [Test async AV insert with good and bad address]...FAIL: fi_eq_readerr index = 0, should be 1
    
    Testing with type = FI_AV_TABLE
    Running av_open_close [Test open and close AVs of varying sizes]...PASS!
    Running av_good_sync [Test sync AV insert with good address]...FAIL: fi_av_insert ret=0, Success
    Running av_null_fi_addr [Test AV insert without specifying fi_addr]...FAIL: fi_av_insert ret=0, Success
    Running av_good_vector_async [Test async AV insert with vector of good addresses]...FAIL: fi_eq_sread ret = -259, Error available
    Running av_zero_async [Test async insert AV insert of zero addresses]...FAIL: fi_eq_sread ret = -259, Error available
    Running av_good_2vector_async [Test async AV inserts with two address vectors]...FAIL: fi_eq_sread ret = -259, Error available
    
    Testing with invalid address
    Running av_bad_sync [Test sync AV insert of bad address]...PASS!
    Running av_goodbad_vector_sync [Test sync AV insert of 1 good and 1 bad address]...FAIL: fi_av_insert ret=0, should be 1
    Running av_goodbad_vector_async [Test async AV insert with good and bad address]...FAIL: count = 0, should be 1
```
I don't know if these are related to the changes here or not.

**Comment by aingerson** [comment]
@shefty Yeah, they are. I'm looking into it. I think the issue is in verbs which calls ofi_str_toaddr on the UD path. I'm looking into it.

**Comment by j-xiong** [comment]
All the failed sub-tests were previous skipped because "test does not yet support FI_FORMAT_UNSPEC". So this is not really a regression.

**Comment by shefty** [comment]
We still need to do something, so that we don't have CI failures forever.  The socket issues are bad enough...

**Comment by j-xiong** [comment]
Ideally we would want the verbs provider to accept sockaddr_in as supported input for fi_av_insert (which currently only accepts FI_ADDR_IB_UD). There can be a lot of work involved.

A quicker approach to address this issue is to let fi_getinfo correctly return -FI_ENODATA when an unsupported address format is requested. The "fi_av_test" test does set "hints->addr_format = FI_SOCKADDR;" before calling fi_getinfo. It passes the check by rxd even if verbs dgram doesn't support it. There must be a missing link somewhere inside rxd or verbs.

**Comment by aingerson** [comment]
I have a quick fix by taking out the redundant call to ofi_str_toaddr and returning the correct format by verbs. There's something funky going on with the verbs UD addressing in the getinfo call. I think more work is needed to get that working properly but this will at least get it working for now.
I can add an extra check for the correct formatting to rxd as well. I'll open a PR in a bit.

**Comment by aingerson** [comment]
@j-xiong @shefty After merging the verbs change, I rebuilt the Intel CI and it's passing now btw

File: src/common.c [hunk 2]
```diff
@@ -538,12 +538,87 @@ static int ofi_str_to_sin6(const char *str, void **addr, size_t *len)
...
 {
 	*addr_format = ofi_addr_format(str);
-	if (*addr_format == FI_FORMAT_UNSPEC)
+	if (*addr_format == FI_FORMAT_UNSPEC) {
+		if (!ofi_hostname_toaddr(str, addr_format, addr, len))
+			return 0;
+		if (!ofi_ifname_toaddr(str, addr_format, addr, len))
+			return 0;
 		return -FI_EINVAL;
+	}
```
Comment (shefty) [inline src/common.c]: Can you remove the if statement and move the code into the switch statement below?

Comment (j-xiong) [reply src/common.c]: Good catch. Done.

**Pull request merged**